graph/cytoscape/cytoscape.go

```diff
@@ -261,41 +254,85 @@ func buildConfig(trafficMap graph.TrafficMap, nodes *[]*NodeWrapper, edges *[]*E
 	}
 }
 
-func addNodeTelemetry(s *graph.Node, nd *NodeData) {
-	httpIn := getRate(s.Metadata, "httpIn")
-
-	if httpIn > 0.0 {
-		nd.HttpIn = fmt.Sprintf("%.2f", httpIn)
-
-		httpIn3xx := getRate(s.Metadata, "httpIn3xx")
-		httpIn4xx := getRate(s.Metadata, "httpIn4xx")
-		httpIn5xx := getRate(s.Metadata, "httpIn5xx")
-
-		if httpIn3xx > 0.0 {
-			nd.HttpIn3xx = fmt.Sprintf("%.2f", httpIn3xx)
-		}
-		if httpIn4xx > 0.0 {
-			nd.HttpIn4xx = fmt.Sprintf("%.2f", httpIn4xx)
+func addNodeTelemetry(n *graph.Node, nd *NodeData) {
+	nd.Traffic = []ProtocolTraffic{}
+	for _, p := range graph.Protocols {
+		protocolTraffic := ProtocolTraffic{Protocol: p.Name}
+		for _, r := range p.NodeRates {
+			if rateVal := getRate(n.Metadata, r.Name); rateVal > 0.0 {
+				if protocolTraffic.Rates == nil {
+					protocolTraffic.Rates = make(map[string]string)
+				}
+				protocolTraffic.Rates[r.Name] = fmt.Sprintf("%.*f", r.Precision, rateVal)
+			}
 		}
-		if httpIn5xx > 0.0 {
-			nd.HttpIn5xx = fmt.Sprintf("%.2f", httpIn5xx)
+		if protocolTraffic.Rates != nil {
+			nd.Traffic = append(nd.Traffic, protocolTraffic)
 		}
 	}
+}
 
-	httpOut := getRate(s.Metadata, "httpOut")
-
-	if httpOut > 0.0 {
-		nd.HttpOut = fmt.Sprintf("%.2f", httpOut)
+func addEdgeTelemetry(e *graph.Edge, ed *EdgeData) {
+	if val, ok := e.Metadata["isMTLS"]; ok {
+		ed.IsMTLS = val.(bool)
 	}
-
-	tcpIn := getRate(s.Metadata, "tcpIn")
-	tcpOut := getRate(s.Metadata, "tcpOut")
-
-	if tcpIn > 0.0 {
-		nd.TcpIn = fmt.Sprintf("%.2f", tcpIn)
+	if val, ok := e.Metadata["responseTime"]; ok {
+		responseTime := val.(float64)
+		ed.ResponseTime = fmt.Sprintf("%.0f", responseTime)
 	}
-	if tcpOut > 0.0 {
-		nd.TcpOut = fmt.Sprintf("%.2f", tcpOut)
+	if val, ok := e.Source.Metadata["isUnused"]; ok {
+		ed.IsUnused = val.(bool)
+	}
+
+	ed.Traffic = []ProtocolTraffic{}
+	for _, p := range graph.Protocols {
+		protocolTraffic := ProtocolTraffic{Protocol: p.Name}
+		total := 0.0
+		err := 0.0
+		var percentErr, percentReq graph.Rate
+		for _, r := range p.EdgeRates {
+			rateVal := getRate(e.Metadata, r.Name)
+			switch {
+			case r.IsTotal:
+				total = rateVal
+			case r.IsErr:
+				err += rateVal
```

Comment:
comments would be nice around here. I think you `+=` because there can be multiple "IsErr" rates defined and thus you need to sum them all up. Total, IsPercentErr, IsPercentReq - those do not get summed via `+=` only because you don't expect more than one of those error, so setting them with `=` rather then summing them with `+=` works.

Reply:
Unless you think it's a blocker I'll add comments here in the follow-up gRPC PR.

Reply:
I added comments in the upcoming gRPC PR.